File: app/src/com/ternaryop/photoshelf/fragment/BirthdaysPublisherFragment.java
package com.ternaryop.photoshelf.fragment;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.os.Bundle;
import android.os.Environment;
import android.util.Pair;
import android.util.SparseBooleanArray;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.GridView;

import com.ternaryop.photoshelf.Constants;
import com.ternaryop.photoshelf.R;
import com.ternaryop.photoshelf.activity.TagPhotoBrowserActivity;
import com.ternaryop.photoshelf.adapter.GridViewPhotoAdapter;
import com.ternaryop.photoshelf.birthday.BirthdayUtils;
import com.ternaryop.photoshelf.db.Birthday;
import com.ternaryop.photoshelf.db.DBHelper;
import com.ternaryop.tumblr.Tumblr;
import com.ternaryop.tumblr.TumblrPhotoPost;
import com.ternaryop.tumblr.TumblrPost;
import com.ternaryop.utils.AbsProgressBarAsyncTask;
import com.ternaryop.utils.ImageUtils;

public class BirthdaysPublisherFragment extends AbsPhotoShelfFragment implements GridView.MultiChoiceModeListener, OnItemClickListener {
    private static final int PICK_IMAGE_REQUEST_CODE = 100;
    private static final String LOADER_PREFIX = "mediumThumb";

    private GridView gridView;
    private GridViewPhotoAdapter gridViewPhotoAdapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_birthdays_publisher, container, false);

        gridViewPhotoAdapter = new GridViewPhotoAdapter(getActivity(), LOADER_PREFIX);
        
        gridView = (GridView)rootView.findViewById(R.id.gridview);
        gridView.setAdapter(gridViewPhotoAdapter);
        gridView.setOnItemClickListener(this);
        gridView.setMultiChoiceModeListener(this);
        
        refresh();
        
        return rootView;
    }

    private void refresh() {
        new AbsProgressBarAsyncTask<Void, Void, List<Pair<Birthday, TumblrPhotoPost>>>(getActivity(), getString(R.string.shaking_images_title)) {

            @Override
            protected List<Pair<Birthday, TumblrPhotoPost>> doInBackground(Void... voidParams) {
                Calendar now = Calendar.getInstance(Locale.US);
                return BirthdayUtils.getPhotoPosts(getActivity(), now);
            }

            @Override
            protected void onPostExecute(List<Pair<Birthday, TumblrPhotoPost>> posts) {
                super.onPostExecute(null);
                if (getError() == null) {
                    gridViewPhotoAdapter.clear();
                    gridViewPhotoAdapter.addAll(posts);
                    gridViewPhotoAdapter.notifyDataSetChanged();
                }
            }
        }.execute();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.birtdays_publisher, menu);
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
        case R.id.action_refresh:
            refresh();
            return true;
        case R.id.action_selectall:
            selectAll(true);
            return true;
        default:
            return super.onOptionsItemSelected(item);
        }
    }
    
    private void selectAll(boolean select) {
        if (select) {
            for (int i = 0; i < gridView.getCount(); i++) {
                gridView.setItemChecked(i, select);
            }
        } else {
            gridView.clearChoices();
        }
    }

    private void publish(final ActionMode mode, final boolean saveAsDraft) {
        new AbsProgressBarAsyncTask<Void, String, List<TumblrPhotoPost>>(getActivity(), "") {
            @Override
            protected void onProgressUpdate(String... values) {
                getProgressDialog().setMessage(values[0]);
            }

            @Override
            protected List<TumblrPhotoPost> doInBackground(Void... voidParams) {
                try {
                    InputStream is = getContext().getAssets().open("cake.png");
                    Bitmap cakeImage = BitmapFactory.decodeStream(is);
                    is.close();

                    for (TumblrPhotoPost post : getCheckedPosts()) {
                        String name = post.getTags().get(0);
                        publishProgress(getContext().getString(R.string.sending_cake_title, name));
                        createBirthdayPost(cakeImage, post, saveAsDraft);
                    }
                } catch (Exception e) {
                    setError(e);
                }
                return null;
            }
            
            protected void onPostExecute(List<TumblrPhotoPost> result) {
                super.onPostExecute(null);
                if (getError() == null) {
                    mode.finish();
                }
            }
        }.execute();
    }

    private void createBirthdayPost(Bitmap cakeImage, TumblrPhotoPost post, boolean saveAsDraft)
            throws IOException {
        String imageUrl = post.getClosestPhotoByWidth(400).getUrl();
        Bitmap image = ImageUtils.readImage(imageUrl);
        
        final int IMAGE_SEPARATOR_HEIGHT = 10;
        int canvasWidth = image.getWidth();
        int canvasHeight = cakeImage.getHeight() + IMAGE_SEPARATOR_HEIGHT + image.getHeight();
        
        Config config = image.getConfig() == null ? Bitmap.Config.ARGB_8888 : image.getConfig();
        Bitmap destBmp = Bitmap.createBitmap(canvasWidth, canvasHeight, config);
        Canvas canvas = new Canvas(destBmp);
        
        canvas.drawBitmap(cakeImage, (image.getWidth() - cakeImage.getWidth()) / 2, 0, null);
        canvas.drawBitmap(image, 0, cakeImage.getHeight() + IMAGE_SEPARATOR_HEIGHT, null);
        File file = saveImage(destBmp, "birth-" + post.getTags().get(0));
        if (saveAsDraft) {
        	Tumblr.getSharedTumblr(getActivity()).draftPhotoPost(getBlogName(),
        			file,
        			getCaption(post),
        			"Birthday, " + post.getTags().get(0));
        } else {
        	Tumblr.getSharedTumblr(getActivity()).publishPhotoPost(getBlogName(),
        			file,
        			getCaption(post),
        			"Birthday, " + post.getTags().get(0));
        }
        file.delete();
    }

    private String getCaption(TumblrPost post) {
        DBHelper dbHelper = DBHelper
                .getInstance(getActivity().getApplicationContext());
        String name = post.getTags().get(0);
        Birthday birthDay = dbHelper
                .getBirthdayDAO()
                .getBirthdayByName(name, getBlogName());
        Calendar birthDate = Calendar.getInstance();
        birthDate.setTime(birthDay.getBirthDate());
        int age = Calendar.getInstance().get(Calendar.YEAR) - birthDate.get(Calendar.YEAR);
        // caption must not be localized
        String caption = "Happy %1$dth Birthday, %2$s!!";
        return String.format(caption, age, name);
    }
    
    private File saveImage(Bitmap image, String fileName) throws IOException {
        File imageFile = new File(Environment.getExternalStoragePublicDirectory(
                Environment.DIRECTORY_PICTURES), fileName + ".png");
        OutputStream fout = new FileOutputStream(imageFile);

        image.compress(Bitmap.CompressFormat.PNG, 100, fout);

        fout.flush();
        fout.close();
        
        return imageFile;
    }
    
    public void onItemClick(AdapterView<?> parent, View v, int position, long id) {
        TumblrPhotoPost post = gridViewPhotoAdapter.getItem(position).second;
        TagPhotoBrowserActivity.startPhotoBrowserActivityForResult(this, getBlogName(),
                post.getTags().get(0),
                PICK_IMAGE_REQUEST_CODE,
                false);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK && requestCode == PICK_IMAGE_REQUEST_CODE) {
            TumblrPhotoPost post = (TumblrPhotoPost) data.getSerializableExtra(Constants.EXTRA_POST);
            gridViewPhotoAdapter.updatePostByTag(post, true);
        }
    }

    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        mode.setTitle(R.string.select_images);
        mode.setSubtitle(getResources().getQuantityString(R.plurals.selected_items, 1, 1));
        MenuInflater inflater = mode.getMenuInflater();
        inflater.inflate(R.menu.birtdays_publisher_action, menu);
        return true;
    }
    
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        return true;
    }

    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        switch (item.getItemId()) {
        case R.id.action_publish:
            publish(mode, false);
            return true;
        case R.id.action_draft:
            publish(mode, true);
            return true;
        default:
            return false;
        }
    }

    private List<TumblrPhotoPost> getCheckedPosts() {
        SparseBooleanArray checkedItemPositions = gridView.getCheckedItemPositions();
        ArrayList<TumblrPhotoPost> list = new ArrayList<TumblrPhotoPost>();
        for (int i = 0; i < checkedItemPositions.size(); i++) {
            int key = checkedItemPositions.keyAt(i);
            if (checkedItemPositions.get(key)) {
                list.add(gridViewPhotoAdapter.getItem(key).second);
            }
        }
        return list;
    }
    
    public void onDestroyActionMode(ActionMode mode) {
    }

    public void onItemCheckedStateChanged(ActionMode mode, int position,
            long id, boolean checked) {
        int selectCount = gridView.getCheckedItemCount();
        mode.setSubtitle(getResources().getQuantityString(R.plurals.selected_items, 1, selectCount));
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/fragment/BirthdaysPublisherFragment.kt
package com.ternaryop.photoshelf.fragment

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.ActionMode
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.ternaryop.photoshelf.EXTRA_POST
import com.ternaryop.photoshelf.R
import com.ternaryop.photoshelf.activity.TagPhotoBrowserActivity
import com.ternaryop.photoshelf.adapter.OnPhotoBrowseClickMultiChoice
import com.ternaryop.photoshelf.adapter.photo.GridViewPhotoAdapter
import com.ternaryop.photoshelf.api.birthday.BirthdayManager
import com.ternaryop.photoshelf.event.BirthdayEvent
import com.ternaryop.photoshelf.service.PublishIntentService
import com.ternaryop.photoshelf.view.AutofitGridLayoutManager
import com.ternaryop.tumblr.TumblrPhotoPost
import com.ternaryop.widget.WaitingResultSwipeRefreshLayout
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.Calendar
import java.util.Locale

private const val PICK_IMAGE_REQUEST_CODE = 100

class BirthdaysPublisherFragment
    : AbsPhotoShelfFragment(),
    SwipeRefreshLayout.OnRefreshListener, OnPhotoBrowseClickMultiChoice, ActionMode.Callback {

    private lateinit var gridViewPhotoAdapter: GridViewPhotoAdapter
    private lateinit var swipeLayout: WaitingResultSwipeRefreshLayout

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_birthdays_publisher, container, false)

        gridViewPhotoAdapter = GridViewPhotoAdapter(context!!)
        gridViewPhotoAdapter.onPhotoBrowseClick = this

        val layout = AutofitGridLayoutManager(context!!,
            resources.getDimension(R.dimen.grid_layout_thumb_width).toInt())
        val gridView = rootView.findViewById<RecyclerView>(R.id.gridview)
        gridView.adapter = gridViewPhotoAdapter
        gridView.setHasFixedSize(true)
        gridView.layoutManager = layout

        swipeLayout = rootView.findViewById(R.id.swipe_container)
        swipeLayout.setOnRefreshListener(this)
        swipeLayout.isRefreshing = true
        refresh()

        setHasOptionsMenu(true)

        return rootView
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }

    private fun refresh() {
        // do not start another refresh if the current one is running
        if (swipeLayout.isWaitingResult) {
            return
        }
        val count = gridViewPhotoAdapter.itemCount
        gridViewPhotoAdapter.clear()
        gridViewPhotoAdapter.notifyItemRangeRemoved(0, count)
        val now = Calendar.getInstance(Locale.US)
        PublishIntentService.startBirthdayListIntent(context!!, now, blogName!!)
        swipeLayout.setRefreshingAndWaitingResult(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.birtdays_publisher, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_refresh -> {
                refresh()
                true
            }
            R.id.action_selectall -> {
                selectAll()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun selectAll() {
        if (actionMode == null) {
            actionMode = activity!!.startActionMode(this)
        }
        gridViewPhotoAdapter.selectAll()
        updateSubTitle()
    }

    private fun publish(mode: ActionMode, publishAsDraft: Boolean) {
        val selectedBirthdays = ArrayList<BirthdayManager.Birthday>()
        val names = ArrayList<String>()

        for (pair in gridViewPhotoAdapter.selectedItems) {
            names.add(pair.name)
            selectedBirthdays.add(pair)
        }

        PublishIntentService.startPublishBirthdayIntent(context!!, selectedBirthdays, blogName!!, publishAsDraft)
        Toast.makeText(context!!,
            getString(R.string.sending_cake_title, TextUtils.join(", ", names)),
            Toast.LENGTH_LONG).show()
        mode.finish()
    }

    override fun onItemClick(position: Int) {
        if (actionMode == null) {
            onThumbnailImageClick(position)
        } else {
            updateSelection(position)
        }
    }

    override fun onTagClick(position: Int, clickedTag: String) {}

    override fun onItemLongClick(position: Int) {
        if (actionMode == null) {
            actionMode = activity!!.startActionMode(this)
        }
        gridViewPhotoAdapter.selection.toggle(position)
    }

    private fun updateSelection(position: Int) {
        val selection = gridViewPhotoAdapter.selection
        selection.toggle(position)
        if (selection.itemCount == 0) {
            actionMode?.finish()
        } else {
            updateSubTitle()
        }
    }

    private fun updateSubTitle() {
        val selection = gridViewPhotoAdapter.selection
        val selectionCount = selection.itemCount
        actionMode?.subtitle = resources.getQuantityString(
                R.plurals.selected_items_total,
                selectionCount,
                selectionCount,
                gridViewPhotoAdapter.itemCount)
    }

    override fun onThumbnailImageClick(position: Int) {
        val birthdate = gridViewPhotoAdapter.getItem(position)
        TagPhotoBrowserActivity.startPhotoBrowserActivityForResult(this, blogName!!,
                birthdate.name,
                PICK_IMAGE_REQUEST_CODE,
                false)
    }

    override fun onOverflowClick(position: Int, view: View) {}

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK && requestCode == PICK_IMAGE_REQUEST_CODE && data != null) {
            val post = data.getSerializableExtra(EXTRA_POST) as TumblrPhotoPost
            gridViewPhotoAdapter.updatePostByTag(post, true)
        }
    }

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        mode.title = getString(R.string.select_images)
        mode.subtitle = resources.getQuantityString(
                R.plurals.selected_items_total,
                1,
                1,
                gridViewPhotoAdapter.itemCount)
        mode.menuInflater.inflate(R.menu.birtdays_publisher_context, menu)
        gridViewPhotoAdapter.isShowButtons = true
        gridViewPhotoAdapter.notifyDataSetChanged()
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        return true
    }

    override fun onDestroyActionMode(mode: ActionMode) {
        this.actionMode = null
        gridViewPhotoAdapter.isShowButtons = false
        gridViewPhotoAdapter.selection.clear()
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_publish -> {
                publish(mode, false)
                true
            }
            R.id.action_draft -> {
                publish(mode, true)
                true
            }
            else -> false
        }
    }

    override fun onRefresh() {
        refresh()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onBirthdayEvent(event: BirthdayEvent) {
        if (event.birthdayResult == null) {
            swipeLayout.setRefreshingAndWaitingResult(false)
            gridViewPhotoAdapter.sort()
            gridViewPhotoAdapter.notifyDataSetChanged()
        } else {
            val position = gridViewPhotoAdapter.itemCount
            event.birthdayResult.birthdates?.let { list ->
                gridViewPhotoAdapter.addAll(list)
                gridViewPhotoAdapter.notifyItemRangeInserted(position, list.size)
            }
        }
    }
}
